processor/k8sprocessor/processor.go

```diff
@@ -122,119 +70,103 @@ func (kp *kubernetesprocessor) Shutdown(context.Context) error {
 	return nil
 }
 
-func (kp *kubernetesprocessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+// ProcessTraces process traces and add k8s metadata using resource IP or incoming IP as pod origin.
+func (kp *kubernetesprocessor) ProcessTraces(ctx context.Context, td pdata.Traces) (pdata.Traces, error) {
 	rss := td.ResourceSpans()
 	for i := 0; i < rss.Len(); i++ {
 		rs := rss.At(i)
 		if rs.IsNil() {
 			continue
 		}
 
-		var podIP string
-		resource := rs.Resource()
+		_ = kp.processResource(ctx, rs.Resource(), false)
+	}
 
-		// check if the application, a collector/agent or a prior processor has already
-		// annotated the batch with IP.
-		if !resource.IsNil() {
-			podIP = kp.k8sIPFromAttributes(resource.Attributes())
-		}
+	return td, nil
+}
 
-		// Check if the receiver detected client IP.
-		if podIP == "" {
-			if c, ok := client.FromContext(ctx); ok {
-				podIP = c.IP
-			}
+// ProcessMetrics process metrics and add k8s metadata using resource IP, hostname or incoming IP as pod origin.
+func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+	rm := md.ResourceMetrics()
+	for i := 0; i < rm.Len(); i++ {
+		ms := rm.At(i)
+		if ms.IsNil() {
+			continue
 		}
 
-		if podIP != "" {
-			if resource.IsNil() {
-				resource.InitEmpty()
-			}
-			resource.Attributes().InsertString(k8sIPLabelName, podIP)
-		}
+		_ = kp.processResource(ctx, ms.Resource(), true)
+	}
 
-		// Don't invoke any k8s client functionality in passthrough mode.
-		// Just tag the IP and forward the batch.
-		if kp.passthroughMode {
-			continue
-		}
+	return md, nil
+}
 
-		// add k8s tags to resource
-		attrsToAdd := kp.getAttributesForPodIP(podIP)
-		if len(attrsToAdd) == 0 {
+// ProcessLogs process metrics and add k8s metadata using resource IP, hostname or incoming IP as pod origin.
+func (kp *kubernetesprocessor) ProcessLogs(ctx context.Context, ld pdata.Logs) (pdata.Logs, error) {
+	lm := ld.ResourceLogs()
+	for i := 0; i < lm.Len(); i++ {
+		ls := lm.At(i)
+		if ls.IsNil() {
 			continue
 		}
 
-		if resource.IsNil() {
-			resource.InitEmpty()
-		}
-
-		attrs := resource.Attributes()
-		for k, v := range attrsToAdd {
-			attrs.InsertString(k, v)
-		}
+		_ = kp.processResource(ctx, ls.Resource(), true)
 	}
 
-	return kp.nextTraceConsumer.ConsumeTraces(ctx, td)
+	return ld, nil
 }
 
-// ConsumeMetrics process metrics and add k8s metadata using resource hostname as pod origin.
-// TODO: Move to internal data model once it's available in contrib.
-func (kp *kubernetesprocessor) ConsumeMetrics(ctx context.Context, metrics pdata.Metrics) error {
-	mds := internaldata.MetricsToOC(metrics)
+func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pdata.Resource, checkHostname bool) error {
```

Comment:
Instead of `checkHostname` argument and then having a condition in the function, can we define something like `type ipExtractor func(Attributes) IP` and then pass either `k8sIPFromAttributes` or `k8sIPFromHostnameAttributes` as the argument? This function can then just call whatever it receives without worrying about the branching logic.